### PR TITLE
Refactor UI and event system

### LINKED ISSUE #112: Refactor UI
Currently, the UI is a singleton that is being used everywhere, this is not very clean.
Possibly thinking about going back to pure Winston.

- [ ] Check other tools for inspiration

### PR DESCRIPTION
**Pull request by dstallenberg**
Closes #112

### CODE REVIEW

File: libraries/core/lib/search/metaheuristics/SearchAlgorithm.ts [hunk 1]
```diff
@@ -106,75 +99,77 @@ export abstract class SearchAlgorithm<T extends Encoding> {
...
+      );
 
       // Start next iteration of the search process
       await this._iterate(budgetManager, terminationManager);
 
       // Inform the budget manager and listeners that an iteration happened
       budgetManager.iteration(this);
 
-      this._eventManager.emitEvent("onSearchIterationComplete");
+      (<TypedEmitter<Events>>process).emit(
```
Comment (mitchellolsthoorn) [inline libraries/core/lib/search/metaheuristics/SearchAlgorithm.ts:158]: We could maybe remove one of these events as we don't need an event as the start and end of a while loop

File: libraries/log/package.json [hunk 2]
```diff
@@ -0,0 +1,53 @@
+{
+  "name": "@syntest/log",
+  "version": "0.1.0",
```
Comment (mitchellolsthoorn) [inline libraries/log/package.json:3]: If we want the version to be 0.1.0 when release, we might need to lower the initial version

**Review by mitchellolsthoorn — CHANGES_REQUESTED** [review]

File: libraries/log/package.json (hunk 2 above)
Comment (dstallenberg) [reply libraries/log/package.json:3]: Fixed.

File: libraries/cli/lib/index.ts [hunk 3]
```diff
@@ -15,15 +15,18 @@
...
+export * from "./module/plugins/PluginType";
+export * from "./module/plugins/ListenerPlugin";
+
 export * from "./module/Command";
 export * from "./module/Module";
 export * from "./module/Plugin";
 export * from "./module/Tool";
 
-export * from "./plugin/UserInterfacePlugin";
+export * from "./userInterfaces/UserInterface";
```
[the remark this answers is not in the document]
Comment (dstallenberg) [reply libraries/cli/lib/index.ts:27]: Fixed. Using dashes now.

File: libraries/cli/lib/ModuleManager.ts [hunk 4]
```diff
@@ -86,17 +88,23 @@ export class ModuleManager {
   }
 
   async prepare() {
+    ModuleManager.LOGGER.info("Preparing modules");
     for (const module of this.modules.values()) {
       if (module.prepare) {
+        ModuleManager.LOGGER.info("Preparing module: " + module.name + "");
```
[the remark this answers is not in the document]
Comment (dstallenberg) [reply libraries/cli/lib/ModuleManager.ts:94]: Fixed.

File: libraries/core/lib/search/metaheuristics/SearchAlgorithm.ts (hunk 1 above)
Comment (dstallenberg) [reply libraries/core/lib/search/metaheuristics/SearchAlgorithm.ts:158]: After a discussion we decided that it cannot hurt to keep both events in here.

File: .github/workflows/pr-test.yml [hunk 5]
```diff
@@ -103,6 +103,7 @@ jobs:
           - libraries/cfg-core
           - libraries/cli
           - libraries/core
+          - libraries/log
```
[the remark this answers is not in the document]
Comment (dstallenberg) [reply .github/workflows/pr-test.yml:106]: Fixed.

File: .github/workflows/main-test.yml [hunk 6]
```diff
@@ -65,6 +65,7 @@ jobs:
           - libraries/cfg-core
           - libraries/cli
           - libraries/core
+          - libraries/log
```
[the remark this answers is not in the document]
Comment (dstallenberg) [reply .github/workflows/main-test.yml:68]: Fixed.

File: libraries/cli/lib/bin.ts [hunk 7]
```diff
@@ -36,35 +35,87 @@ async function main() {
...
+    }
+    LOGGER.info("Cleaning up...");
     ModuleManager.instance.cleanup();
+    LOGGER.info("Cleanup done! Exiting...");
   });
 
-  // execute program
+  userInterface.printHeader("Modules loaded:");
+
+  for (const module of ModuleManager.instance.modules.values()) {
```
[the remark this answers is not in the document]
Comment (dstallenberg) [reply libraries/cli/lib/bin.ts]: Fixed.

**Review by mitchellolsthoorn — APPROVED** [review]

**Pull request merged**